packages/ckeditor5-ui/src/textarea/textareaview.ts

```diff
@@ -22,24 +22,136 @@ export default class TextareaView extends InputBase<HTMLTextAreaElement> {
 	 * @observable
 	 * @default 1
 	 */
-	declare public rows: number;
+	declare public minRows: number;
+
+	/**
+	 * Specifies the maximum number of rows.
+	 *
+	 * @observable
+	 * @default 5
+	 */
+	declare public maxRows: number;
+
+	/**
+	 * TODO
+	 *
+	 * @observable
+	 * @default null
+	 */
+	declare public height: number | null;
+
+	/**
+	 * TODO
+	 *
+	 * @observable
+	 * @default 'both'
+	 */
+	declare public resize: 'both' | 'horizontal' | 'vertical' | 'none';
 
 	/**
 	 * @inheritDoc
 	 */
 	constructor( locale?: Locale ) {
 		super( locale );
 
-		this.set( 'rows', 1 );
+		const toPx = toUnit( 'px' );
+
+		this.set( 'minRows', 1 );
+		this.set( 'maxRows', 5 );
+		this.set( 'height', null );
+		this.set( 'resize', 'both' );
 
 		const bind = this.bindTemplate;
 
 		this.template!.tag = 'textarea';
 
 		this.extendTemplate( {
 			attributes: {
-				rows: bind.to( 'rows' )
+				style: {
+					height: bind.to( 'height', height => height ? toPx( height ) : null ),
+					resize: bind.to( 'resize' )
+				},
+				rows: bind.to( 'minRows' )
 			}
 		} );
+
+		this.on( 'input', () => {
+			this._autoGrow( true );
+		} );
+
+		this.on( 'change:value', () => {
+			// The content needs to be updated by the browser after the value is changed. It takes a few ms.
+			setTimeout( () => this._autoGrow(), 50 );
+		} );
+	}
+
+	/**
+	 * TODO
+	 */
+	public override reset(): void {
+		super.reset();
+
+		this._autoGrow();
 	}
+
+	/**
+	 * TODO
+	 */
+	private _autoGrow( shouldScroll?: boolean ): void {
+		const viewElement = this.element!;
+		const singleLineContentClone = getElementClone( viewElement, '1' );
+		const fullTextValueClone = getElementClone( viewElement, viewElement.value );
+		const singleLineContentStyles = singleLineContentClone.ownerDocument.defaultView!.getComputedStyle( singleLineContentClone );
+
+		const verticalPaddings = parseFloat( singleLineContentStyles.paddingTop ) + parseFloat( singleLineContentStyles.paddingBottom );
+		const borders = getBorderWidths( singleLineContentClone );
+		const lineHeight = parseFloat( singleLineContentStyles.lineHeight );
+		const verticalBorder = borders.top + borders.bottom;
+
+		const singleLineAreaDefaultHeight = new Rect( singleLineContentClone ).height;
+		const numberOfLines = Math.round( ( fullTextValueClone.scrollHeight - verticalPaddings ) / lineHeight );
+
+		const maxHeight = this.maxRows * lineHeight + verticalPaddings + verticalBorder;
```

Comment:
Do we really need this height calculation?

Reply:
Yes and no. Using `height` instead of manipulating `rows` gives us an advantage when the element is in some layout (flex or grid). The thing is the layout ignores `rows`, so even if we want a specific max number of rows, flex/grid will stretch or squeeze textarea to fit the layout. With a specific `height`, it's no longer the case. That's why I think this is more bulletproof.